crates/language/src/language_registry.rs

```diff
@@ -709,21 +706,44 @@ impl LanguageRegistry {
             })
         });
         self.find_matching_language(move |language_name, config, current_best_match| {
+            let filename = filename?;
+
+            let matching_suffix_len = |acc: usize, suffix: &String| {
+                let ext = ".".to_string() + suffix;
+                if filename.ends_with(&ext) || filename == suffix {
+                    acc.max(suffix.len())
+                } else {
+                    acc
+                }
+            };
+
             let path_matches_default_suffix = || {
-                config
-                    .path_suffixes
-                    .iter()
-                    .any(|suffix| path_suffixes.contains(&Some(suffix.as_str())))
+                let len = config.path_suffixes.iter().fold(0, matching_suffix_len);
```

Comment:
I suppose that this could be short circuited: if any of the path suffixes registered for a lang match, perhaps we can assume that the language doesn't also include another longer match later in the list? Most languages don't register too many suffixes, but some do: `bash` has 19, C++ has 15, JS and TS both have 4, markdown has 5, etc. C, Go, Rust, PHP only have 1, fwiw.